### PR TITLE
Add draft unit and integration tests, code coverage is around 75%

### LINKED ISSUE #14: Implement integration tests
tests for:

- creating connection
- tracker
- tracker + x2 nodes + publisher

### PR DESCRIPTION
**Pull request by mirpo**
fixes #14

### CODE REVIEW

File: publisher-libp2p.js [hunk 1]
```diff
@@ -15,7 +15,7 @@ connection.once('node:ready', () => {
     setInterval(() => {
         const msg = `Hello world, ${new Date().toLocaleString()}`
 
-        if (publisher.connection.isReady()) {
+        if (publisher.connection.isStarted()) {
```
Comment (harbu) [inline publisher-libp2p.js]: Need to use new async connection here.

File: publisher.js [hunk 2]
```diff
@@ -15,7 +15,7 @@ connection.once('node:ready', () => {
     setInterval(() => {
         const msg = 'Hello world, ' + new Date().toLocaleString()
 
-        if (publisher.connection.isReady()) {
+        if (publisher.connection.isStarted()) {
```
Comment (harbu) [inline publisher.js]: Need to use new async connection here.

File: src/connection/Connection.js [hunk 3]
```diff
@@ -58,13 +57,12 @@ module.exports = class Connection extends EventEmitter {
...
 
             debug('node has started: %s', this.node.isStarted())
-            this.isStarted = this.node.isStarted()
 
             this.emit(events.READY)
         })
     }
 
-    isReady() {
+    isStarted() {
```
Comment (harbu) [inline src/connection/Connection.js]: I think this can be removed.

File: src/protocol/TrackerNode.js [hunk 4]
```diff
@@ -22,10 +22,14 @@ module.exports = class TrackerNode extends EventEmitter {
 
     async onConnectToTracker(tracker) {
         if (isTracker(getAddress(tracker)) && !this.connection.isConnected(tracker)) {
-            await this.connection.connect(tracker)
-
-            this.tracker = tracker
-            this.emit(events.CONNECTED_TO_TRACKER, tracker)
+            await this.connection.connect(tracker).catch((err) => {
```
Comment (harbu) [inline src/protocol/TrackerNode.js:34]: Does catch also get invoked when `resolve` is invoked inside Promise? I thought it only gets invoked when `reject` is invoked?

File: test/integration/connection.test.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
+const assert = require('assert')
+const Connection = require('../../src/connection/Connection')
+
+jest.setTimeout(30000)
+
+describe('create two connections and init connection between them', () => {
+    it('should be able to start and stop successfully', (done) => {
+        const connection = new Connection('127.0.0.1', 30337)
```
Comment (harbu) [inline test/integration/connection.test.js]: async connection creation

File: test/integration/publisher.test.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+const Connection = require('../../src/connection/Connection')
+const Node = require('../../src/logic/Node')
+const Publisher = require('../../src/logic/Publisher')
+
+jest.setTimeout(30000)
+
+describe('publisher and node connection', () => {
+    it('should be able to start publisher and node, send message, recieve and then stop successfully', (done) => {
+        // create node
+        const connection = new Connection('127.0.0.1', 30333, '', true)
```
Comment (harbu) [inline test/integration/publisher.test.js]: async connection

File: test/integration/tracker.test.js [hunk 7]
```diff
@@ -1,40 +1,37 @@
...
+        // create tracker
+        const connection = new Connection('127.0.0.1', 30300, PRIVATE_KEY)
```
Comment (harbu) [inline test/integration/tracker.test.js]: async connection

File: test/unit/connection.test.js [hunk 8]
```diff
@@ -0,0 +1,14 @@
+const assert = require('assert')
+const Connection = require('../../src/connection/Connection')
+
+describe('create connection', () => {
+    it('should be able to start and stop successfully', (done) => {
+        const connection = new Connection('127.0.0.1', 30333)
+
+        connection.on('node:ready', () => {
```
Comment (harbu) [inline test/unit/connection.test.js]: async connection

File: test/unit/encoder.test.js [hunk 9]
```diff
@@ -0,0 +1,40 @@
+const assert = require('assert')
+const encoder = require('../../src/helpers/MessageEncoder')
+const Connection = require('../../src/connection/Connection')
+
+describe('encoder', () => {
+    it('check all codes', (done) => {
+        assert.equal(encoder.STATUS, 0)
```
Comment (harbu) [inline test/unit/encoder.test.js:7]: IMO this level of testing is not necessary. It is obvious from code what the values are.

File: test/unit/encoder.test.js [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+        assert.equal(encoder.getMsgPrefix(encoder.PEERS), 'PEERS')
+        assert.equal(encoder.getMsgPrefix(encoder.SUBSCRIBE), 'SUBSCRIBE')
+        assert.equal(encoder.getMsgPrefix(encoder.PUBLISH), 'PUBLISH')
+        assert.equal(encoder.getMsgPrefix(encoder.STREAM), 'STREAM')
+
+        done()
+    })
+
+    it('check streamMessage encoding/decoding', (done) => {
+        const json = encoder.streamMessage('stream-address', 'node-address')
```
Comment (harbu) [inline test/unit/encoder.test.js:27]: Can we use `stream-id` instead of `stream-address`.

**Review by harbu — CHANGES_REQUESTED** [review]
Looking good! Couple changes.

* Need async connections
* Multiple ports  for trackers so can be run concurrently
* Disconnection issue?

File: src/connection/Connection.js [hunk 11]
```diff
@@ -146,6 +101,10 @@ class Connection extends EventEmitter {
         return new Promise((resolve, reject) => this.node.dial(address, (err, peerInfo) => (err ? reject(err) : resolve(peerInfo))))
     }
 
+    async stop(callback = true) {
+        return new Promise((resolve, reject) => this.node.stop((err) => (err ? reject(err) : resolve(callback))))
```
Comment (harbu) [inline src/connection/Connection.js:105]: Good one! I actually need this in broker. 💯

File: test/unit/encoder.test.js (hunk 10 above)
Comment (mirpo) [reply test/unit/encoder.test.js:27]: Sure

**Comment by harbu** [comment]
LGTM, but we should make Travis pass. Currently not liking ESLint 😆

**Review by harbu — APPROVED** [review]

**Pull request merged**